### ISSUE TITLE
 "Error importing" when loading job files from disk 

### ISSUE THREAD

**Issue by labrys** [sabnzbd/sabnzbd#952]
I get a lot of "Error importing" errors.
![image](https://user-images.githubusercontent.com/1867464/27456246-5bd7f53c-576e-11e7-8f61-e41e178d41b9.png)
Is it possible to give a more descriptive error message for these?
I notice in the log it reads:
```
2017-06-22 17:15:08,460::ERROR::[nzbstuff:1342] Error importing <NzbFile: filename=4f3a3eec98984e2b9302b60767c268ba.vol003+04.par2, type=None>
```
Would it be possible to get a more descriptive message like that in the UI?  Preferably with the name of the Queue item.

**Comment by Safihre** [comment]
I can make better error messages, but we should also try to find the reason you are getting these problems. 
Does it show in the log the trackback?

**Comment by Safihre** [comment]
The problem was that the `<>` were seen as HTML. Of course all HTML should be stripped from the warnings, which is done now.

But when did this error happen? The location where the error is thrown is surprising and will not have logged a traceback.
Did this happen more often?

**Comment by labrys** [comment]
It happens really frequently.  I assumed they were malformed nzb files so hadn't worried about it.

**Comment by Safihre** [comment]
Are these jobs that were already in the queue before the update to 2.2.0 and the rebuild of the queue?

**Comment by labrys** [comment]
No.  I cleared my queue to update to 2.2.0.  Also it happened regularly both before and after the queue changes.

**Comment by Safihre** [comment]
Is it possible for you to share some of the NZB's on which it happens?

**Comment by labrys** [comment]
Sure next one I get I'll post it.  I recently purged my queue so I don't have any right at the moment.  Most likely I'll have some during the next backlog search.

**Comment by Safihre** [comment]
Found another possible reason. Maybe do another pull :) It's a simple update, no queue updates etc this time ;)

**Comment by labrys** [comment]
Already did :).  Also it seems to happen more often when I sort queue in the UI.  I don't know if thats when i happen to notice it because it pushes the results to the top or because of an issue with the sorting.  I used Medusa's manual snatch feature to re-add an NZB that caused the error to try to reproduce it and it completed without issues, again leading me to believe its something with the sorting.  Are files being modified during a sort? if so maybe its a race condition.

**Comment by Safihre** [comment]
So it still happens with the commit I pushed 3 hours ago?

**Comment by labrys** [comment]
No havent had it with the latest, but it usually takes a while to happen. That comment was related to the previous commit.

**Comment by labrys** [comment]
Just happened again.  Once again it was while sorting the Queue (Size asc).
![image](https://user-images.githubusercontent.com/1867464/27626605-ad4181c6-5bb6-11e7-81da-f423cfb4fe34.png)

**Comment by labrys** [comment]
Re-added through manual snatch and moving it to top of Queue I get the normal "missing articles"

**Comment by Safihre** [comment]
Damn. I just pushed a new commit that will tell us more why it is throwing the error so we know what the problem is. Either the file is missing, or the file was empty.
Could you do a pull and let us know which errors you get? I'm not able to reproduce it on my setup yet.

**Comment by labrys** [comment]
Pulled latest.  Will let you know.

**Comment by labrys** [comment]
It looks like the path does not exist:

![image](https://user-images.githubusercontent.com/1867464/27666789-51192848-5c44-11e7-9867-e23c186a4526.png)

**Comment by thezoggy** [comment]
can you provide the nzb for that ?

**Comment by labrys** [comment]
https://gist.github.com/labrys/30ba992fdaab252672331da57b7e1eb8

**Comment by labrys** [comment]
The reason I mentioned the sorting previously though is that the same NZB re-added will usually download (as was the case with this one).  It also occurs most often when sorting the Queue.  Typical use-case is after a backlog search when many files are added to SAB at once, I set all files to the same priority (usually low) and then sort by file size ascending.  This way all the smaller files are downloaded first to reduce total number of items in the queue.  Sometimes I may repeat this process more than once (for example if the backlog is still in progress and more nzbs are snatched or if i want to prioritize a subset of the files).

**Comment by Safihre** [comment]
@thezoggy it's not the NZB, I can also randomly reproduce it (see below, but the different type of problem).
It seems to be something to do with the threading locks I removed/changed, have to dig in to see which one and why.
![image](https://user-images.githubusercontent.com/5703454/27752433-9ecfcbd6-5de9-11e7-8a89-9941e53c6f0f.png)

**Comment by labrys** [comment]
@Safihre If it helps it's been happening for quite a while.... I would guess since roughly around the time that you removed the option for a secondary gui as that's when I started messing around with the sorting more due to the differences in how the different gui's behave.

**Comment by Safihre** [comment]
Really, since then? Oke then I'm shooting in the dark and have to look for higher level problems.

**Comment by labrys** [comment]
If not right then, soon after.  Its been quite a while.

**Comment by Safihre** [comment]
@labrys Are you still experiencing this? I cannot reproduce it anymore..

**Comment by labrys** [comment]
I'm still running 336b1ddba3a922b15724326e51f676854a5d537d. I'll pull the latest and let you know.

**Comment by labrys** [comment]
Just got this error today on commit 
```

**Comment by Safihre** [comment]
@labrys My current hypothesis is that the deletion of a job (from History or Queue, not sure yet) might lead to the deletion of another job with a very similar name.
Do you keep logging on Debug? Because in case you get this error we should check in the logs if maybe the article database was deleted previously.

**Comment by thezoggy** [comment]
Wonder if it's the issue I've seen where when you tell Sab to delete a job it will nuke similarly named jobs (app grabs different release but same release name) so the job is being deleted as it's imported...

**Comment by Safihre** [comment]
@thezoggy yes I think so, I also had this but never was able to  reproduce it reliably.

**Comment by labrys** [comment]
@thezoggy I do have debug logging enabled.  If it helps any I usually get  along string of these.  Also possibly related i get occaisonal errors about renames .. I'd have to get the log for it next time but it ends in /__admin__/__renames__ or something like that.

**Comment by Safihre** [comment]
Yes that could also be part of it. When such an error happens, or the import error, we should check the logs to see if it was previously erroneously  deleted.

**Comment by Safihre** [comment]
@labrys Still experiencing this problem? 
Just checking in, always hoping one of my other changes magically fixed it 🤣

**Comment by labrys** [comment]
I haven't updated recently.  I'll update today and let you know.

**Comment by Safihre** [comment]
@labrys and @thezoggy: I have added extra logging that will log any file or directory removal occurring and the NZF-ID.
In case this happens again (or to anyone having the same problem), can you search in all logs in your [logs directory](https://sabnzbd.org/wiki/advanced/directory-setup) for the lines that contain the NZF-ID? (`SABnbzd_nzf_xxxxxxx`)?

If it is not logged as being deleted somewhere else, then it means that some external program is deleting the files, just like in #1047.
I will admit that I would have expected many more bug reports over the past 4 months.

**Comment by labrys** [comment]
The error importing lines reference the NzbFile filename but not the NZF-ID.  Is there an easy way to determine the NZF-ID from the filename?

**Comment by labrys** [comment]
I think this is one of them:
```
2017-10-11 00:13:14,900::DEBUG::[__init__:913] Loading data for SABnzbd_nzo_dcufan from H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__\SABnzbd_nzo_dcufan
2017-10-11 00:23:48,525::INFO::[nzbqueue:527] Resumed nzo: SABnzbd_nzo_dcufan
2017-10-11 00:23:48,858::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 00:24:27,543::DEBUG::[misc:1088] Removing file \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__\SABnzbd_nzo_dcufan
2017-10-11 00:34:56,289::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 00:35:01,332::DEBUG::[__init__:913] Loading data for SABnzbd_nzo_dcufan from H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__\SABnzbd_nzo_dcufan
2017-10-11 00:37:25,155::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 00:37:30,308::DEBUG::[__init__:913] Loading data for SABnzbd_nzo_dcufan from H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__\SABnzbd_nzo_dcufan
2017-10-11 00:40:29,102::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 00:40:34,312::DEBUG::[__init__:913] Loading data for SABnzbd_nzo_dcufan from H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__\SABnzbd_nzo_dcufan
2017-10-11 00:41:00,551::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 01:02:39,378::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 01:39:41,934::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 06:16:24,865::INFO::[nzbqueue:527] Resumed nzo: SABnzbd_nzo_dcufan
2017-10-11 06:16:43,117::DEBUG::[__init__:877] Saving data for SABnzbd_nzo_dcufan in \\?\H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__
2017-10-11 06:16:50,170::DEBUG::[__init__:913] Loading data for SABnzbd_nzo_dcufan from H:\tmp\nzb\downloads\Brooklyn.Nine-Nine.S05E03.720p.HDTV.x264-AVS\__ADMIN__\SABnzbd_nzo_dcufan
```
It looks odd to me that it resumes the nzo, saves data and then removes the files, then saves the data and loads again...

**Comment by Safihre** [comment]
I just added the commit that now also shows the NZF-ID in the warning, so you'll have to do a pull first.

The `nzo` files so far have been fine, it's really the `nzf` files that disappear.

**Comment by shypike** [comment]
@Safihre Are the actual NZF files disappearing?

**Comment by Safihre** [comment]
Yes they are being removed at some other stage. I've seen various places, sometimes during readd of all orphaned jobs (now locked),  when deleting a failed job from history (no idea why) or having 2 jobs being added at same time with same name (also locked now). But so far we don't have enough logging of the original events.
All file deleting from within SABnzbd (and the functions that call the deleting) are now being logged.
In case it happens again and logging was enabled, we'll catch it :)

**Comment by labrys** [comment]
I thought I had posted my logs showing the NZF-ID, but I don't see it here.... Yes it has continued to happen I'll post as soon as the next one occurs.

**Comment by Safihre** [comment]
If you could update to the latest stuff on `develop`, that would be great. It includes the logging we need.

**Comment by Safihre** [comment]
Hmmm, I wrote in the changelog of 2.3.1 to report any `Error Importing` here, but so far no reports anymore. I know not many people read changelogs, but I would have expected at least 1 or 2 reports on the forum.
@labrys @thezoggy Are you still getting this? Now we have the right logging we can maybe find the root.

**Comment by thezoggy** [comment]
I havent seen it recently.. but I've not really been looking for it (never happened to me, but to a friend). I'll keep an eye out on his install this week.

**Comment by Safihre** [comment]
@labrys Hoping you'll tell me you're not getting this error anymore! 🥇 😇

**Comment by labrys** [comment]
I saw it once about a week ago but I haven't updated since before thanksgiving either.

**Comment by labrys** [comment]
Sorry to say:
![image](https://user-images.githubusercontent.com/1867464/33796292-4a8f751c-dcbf-11e7-9161-bb2024bfb2cf.png)

```
sabnzbd.log:2017-12-09 04:11:18,924::DEBUG::[__init__:877] [sabnzbd.nzbstuff.__init__] Saving data for SABnzbd_nzf_jb55x8 in \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__
sabnzbd.log:2017-12-09 04:11:18,930::INFO::[nzbstuff:533] File fbea86cc298a48d1b8ae4c282b5e87be.vol063+56.par2 - SABnzbd_nzf_jb55x8 added to queue
sabnzbd.log:2017-12-09 08:47:34,247::DEBUG::[misc:1184] [sabnzbd.misc.remove_all] Deleting file \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__\SABnzbd_nzf_jb55x8
sabnzbd.log:2017-12-09 08:50:03,802::INFO::[__init__:909] [sabnzbd.nzbstuff.finish_import] \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__\SABnzbd_nzf_jb55x8 missing
sabnzbd.log:2017-12-09 08:50:03,808::WARNING::[nzbstuff:308] Article DB file not found SABnzbd_nzf_jb55x8: <NzbFile: filename=fbea86cc298a48d1b8ae4c282b5e87be.vol063+56.par2, type=None>
```
```
sabnzbd.log:2017-12-09 04:11:18,913::DEBUG::[__init__:877] [sabnzbd.nzbstuff.__init__] Saving data for SABnzbd_nzf_wkvmop in \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__
sabnzbd.log:2017-12-09 04:11:18,918::INFO::[nzbstuff:533] File fbea86cc298a48d1b8ae4c282b5e87be.vol031+32.par2 - SABnzbd_nzf_wkvmop added to queue
sabnzbd.log:2017-12-09 08:47:34,446::DEBUG::[misc:1184] [sabnzbd.misc.remove_all] Deleting file \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__\SABnzbd_nzf_wkvmop
sabnzbd.log:2017-12-09 08:50:03,782::INFO::[__init__:909] [sabnzbd.nzbstuff.finish_import] \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__\SABnzbd_nzf_wkvmop missing
sabnzbd.log:2017-12-09 08:50:03,788::WARNING::[nzbstuff:308] Article DB file not found SABnzbd_nzf_wkvmop: <NzbFile: filename=fbea86cc298a48d1b8ae4c282b5e87be.vol031+32.par2, type=None>
```
```
2017-12-09 04:11:18,826::DEBUG::[__init__:877] [sabnzbd.nzbstuff.__init__] Saving data for SABnzbd_nzf_mtign4 in \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__
2017-12-09 04:11:18,832::INFO::[nzbstuff:533] File fbea86cc298a48d1b8ae4c282b5e87be.part58.rar - SABnzbd_nzf_mtign4 added to queue
2017-12-09 08:47:34,346::DEBUG::[misc:1184] [sabnzbd.misc.remove_all] Deleting file \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__\SABnzbd_nzf_mtign4
2017-12-09 08:50:03,638::INFO::[__init__:909] [sabnzbd.nzbstuff.finish_import] \\?\H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT\__ADMIN__\SABnzbd_nzf_mtign4 missing
2017-12-09 08:50:03,644::WARNING::[nzbstuff:308] Article DB file not found SABnzbd_nzf_mtign4: <NzbFile: filename=fbea86cc298a48d1b8ae4c282b5e87be.part58.rar, type=None>
```

**Comment by Safihre** [comment]
Interesting, so it's `remove_all`. Could you email me the whole log? 
Then I can see what is happening around these calls safihre@sabnzbd.org

**Comment by labrys** [comment]
Sent

**Comment by Safihre** [comment]
Oke I found it: A new job is added in the same folder as a previously failed job. At some point `Delete all failed jobs` is clicked and the new job data gets removed as if it is part of the failed job.
However, I don't think this is due to a bug in SABnzbd. Because when a new NZB is added, it will create a unique folder.
I am suspecting your PP-script `nzbToSickBeard` to automatically delete the job-folder for failed jobs, is this the case?

**Comment by labrys** [comment]
Actually I believe it happens when I use the purge history button on the UI.

**Comment by labrys** [comment]
I don't think `nzbToSickebeard` has any effect on this, as it has no knowledge of the location of the `Incomplete Folder`.  Additionally it only happens when I use the UI to clear the history and there are items in the Queue being downloaded.

**Comment by Safihre** [comment]
Yes it does, because you have disabled `Post process only verified jobs' and so the path is passed to the script even for failed jobs as you can see here:
```
2017-12-08 23:18:55,313::DEBUG::[nzbstuff:1166] Abort job "Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDTV.x264-AVS", due to impossibility to complete it
2017-12-08 23:18:55,480::INFO::[newsunpack:172] Running external script H:\usr\local\nzbtomedia\nzbToSickBeard.py(H:\tmp\nzb\downloads\Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDT, Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDTV.x264-AVS.nzb, Marvels.Agents.of.S.H.I.E.L.D.S05E03.720p.HDTV.x264-AVS, , tv, alt.binaries.tv, 3, https://dognzb.cr/fail/3cd75e77d84d2835408cc8a602fe511a/f0bf2ab430780ffb019151c8c8fe7a89)
```

Let me try to explain again:

1. You download the job `job1` the first time, it's in folder `incomplete/job1`.
2. It fails.
3. The script runs, it deletes  `incomplete/job1`.
4. SABnzbd adds it to the history database and saves the path in the database `incomplete/job1`
5. New `job1` gets added, it gets stored in `incomplete/job1` <-- This would never have happened if the script did not delete `incomplete/job1` in step 3.
6. You run `Purge history`, SABnzbd checks in the history database and finds the old `job1` with path `incomplete/job1`. It goes on an deletes `incomplete/job1`.

Step 6 would never resulted in problems if it wasn't for step 3.

**Comment by labrys** [comment]
@Safihre Also I'm fairly familiar with the `nzbToSickbeard` code as a I did some major overhauling of it in prep for py3 compatibility, so it's not an assumption that it doesn't know the location of the `Incomplete Folder`, the only folders it knows of are the `default_downloadDirectory` for uncategorized downloads which it does not process and the folder passed to it by the client (`Medusa` in my case) which is the `series` category folder.
![image](https://user-images.githubusercontent.com/1867464/33797102-f16a9898-dccf-11e7-8756-eab6a049cdb2.png)

**Comment by labrys** [comment]
I'll look at the Medusa code as it was rewritten a while back for nzb processing by another dev to see if its passing the incomplete folder instead of the intended processing folder.  I assume the sab api can expose the incomplete folder location as the only folder I have configured in Medusa is the post processing directory.

**Comment by Safihre** [comment]
This is the `nzbToSickbeard` of `nzbToMedia`, right? 
Here I see that it does pass the path given by SAB:
https://github.com/clinton-hall/nzbToMedia/blob/master/nzbToMedia.py#L734-L747
Which calls `cleanDir` at the end, that removes all folders etc? 
https://github.com/clinton-hall/nzbToMedia/blob/master/core/nzbToMediaUtil.py#L736-L761

It's also that we haven't had any other reports on the forum/Reddit for months, making it more and more likely to be an external factor.

**Comment by Safihre** [comment]
Closing this, I am 99.99% sure now that SABnzbd is doing the correct thing and an external tool deleted the incomplete job-folder. I tested it now a couple of times and this the only way to reproduce this error.
